File: uf-client-service/src/main/java/com/kynetics/uf/android/UpdateFactoryService.java
/*
 * Copyright © 2017-2018  Kynetics  LLC
 *
 *  All rights reserved. This program and the accompanying materials
 *  are made available under the terms of the Eclipse Public License v1.0
 *  which accompanies this distribution, and is available at
 *  http://www.eclipse.org/legal/epl-v10.html
 */

package com.kynetics.uf.android;

import android.app.Service;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.os.Handler;
import android.os.IBinder;
import android.os.Message;
import android.os.Messenger;
import android.os.RemoteException;
import android.support.annotation.NonNull;
import android.util.Log;
import android.widget.Toast;

import com.kynetics.uf.android.api.UFServiceConfiguration;
import com.kynetics.uf.android.api.UFServiceMessage;
import com.kynetics.uf.android.api.UFServiceMessage.Suspend;
import com.kynetics.uf.android.content.SharedPreferencesWithObject;
import com.kynetics.uf.android.ui.MainActivity;
import com.kynetics.updatefactory.ddiclient.api.ClientBuilder;
import com.kynetics.updatefactory.ddiclient.api.api.DdiRestApi;
import com.kynetics.updatefactory.ddiclient.core.UFService;
import com.kynetics.updatefactory.ddiclient.core.model.Event;
import com.kynetics.updatefactory.ddiclient.core.model.State;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.Observable;
import java.util.Observer;

import okhttp3.OkHttpClient;

import static android.content.Intent.FLAG_ACTIVITY_NEW_TASK;
import static com.kynetics.uf.android.api.UFServiceCommunicationConstants.MSG_AUTHORIZATION_REQUEST;
import static com.kynetics.uf.android.api.UFServiceCommunicationConstants.MSG_AUTHORIZATION_RESPONSE;
import static com.kynetics.uf.android.api.UFServiceCommunicationConstants.MSG_CONFIGURE_SERVICE;
import static com.kynetics.uf.android.api.UFServiceCommunicationConstants.MSG_REGISTER_CLIENT;
import static com.kynetics.uf.android.api.UFServiceCommunicationConstants.MSG_RESUME_SUSPEND_UPGRADE;
import static com.kynetics.uf.android.api.UFServiceCommunicationConstants.MSG_SEND_STRING;
import static com.kynetics.uf.android.api.UFServiceCommunicationConstants.MSG_SERVICE_CONFIGURATION_STATUS;
import static com.kynetics.uf.android.api.UFServiceCommunicationConstants.MSG_SYNCH_REQUEST;
import static com.kynetics.uf.android.api.UFServiceCommunicationConstants.MSG_UNREGISTER_CLIENT;
import static com.kynetics.uf.android.api.UFServiceCommunicationConstants.SERVICE_DATA_KEY;
import static com.kynetics.uf.android.api.UFServiceMessage.Suspend.DOWNLOAD;
import static com.kynetics.uf.android.api.UFServiceMessage.Suspend.NONE;
import static com.kynetics.uf.android.api.UFServiceMessage.Suspend.UPDATE;

/**
 * @author Daniele Sergio
 */
public class UpdateFactoryService extends Service implements UpdateFactoryServiceCommand {
    private static final String TAG = UpdateFactoryService.class.getSimpleName();

    public static UpdateFactoryServiceCommand getUFServiceCommand(){
        return ufServiceCommand;
    }

    @Override
    public void authorizationGranted() {
        ufService.setAuthorized(true);
    }

    @Override
    public void authorizationDenied() {
        ufService.setAuthorized(false);
    }


    @Override
    public void configureService() {
        if(ufService!=null){
            ufService.stop();
        }
        buildServiceFromPreferences(true); // TODO: 11/14/17 fix workaround
    }

    @Override
    public void onCreate() {
        super.onCreate();
        initSharedPreferencesKeys();
        ufServiceCommand = this;
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        final Serializable serializable = intent.getSerializableExtra(SERVICE_DATA_KEY);
        boolean startNewService = false;
        if(serializable instanceof UFServiceConfiguration){
            final UFServiceConfiguration serviceConfiguration = (UFServiceConfiguration) serializable;
            saveServiceConfigurationToSharedPreferences(serviceConfiguration);
            startNewService = true;
        }
        buildServiceFromPreferences(startNewService);
        return START_STICKY;
    }

    private OkHttpClient.Builder buildOkHttpClient() {
        return new OkHttpClient.Builder();
    }

    private State getInitialState(boolean startNewService,
                                  SharedPreferencesWithObject sharedPreferences) {
        final Long updatePendingId = UpdateSystem.getUpdatePendingId();
        if(updatePendingId != null){
            Log.e(TAG, "updatePendingId found: " + updatePendingId);
            return new State.UpdateStartedState(updatePendingId);
        }
        return startNewService ? new State.WaitingState(0, null) : sharedPreferences.getObject(sharedPreferencesCurrentStateKey, State.class, new State.WaitingState(0, null));
    }

    private void buildServiceFromPreferences(boolean startNewService) {
        startStopService(false);
        final SharedPreferencesWithObject sharedPreferences = getSharedPreferences(sharedPreferencesFile, MODE_PRIVATE);
        final boolean serviceIsEnable = sharedPreferences.getBoolean(sharedPreferencesServiceEnableKey, false);
        if(serviceIsEnable) {
            final String url = sharedPreferences.getString(sharedPreferencesServerUrlKey, "");
            final String controllerId = sharedPreferences.getString(sharedPreferencesControllerIdKey, "");
            final String gatewayToken = sharedPreferences.getString(sharedPreferencesGatewayToken, "");
            final String targetToken = sharedPreferences.getString(sharedPreferencesTargetToken, "");
            final String tenant = sharedPreferences.getString(sharedPreferencesTenantKey, "");
            final long delay = sharedPreferences.getLong(sharedPreferencesRetryDelayKey, 30000);
            final State initialState = getInitialState(startNewService, sharedPreferences);
            final boolean apiMode = sharedPreferences.getBoolean(sharedPreferencesApiModeKey, true);
            final HashMap<String,String> defaultArgs = new HashMap<>();
            final Map<String,String> args = sharedPreferences.getObject(sharedPreferencesArgs, defaultArgs.getClass());
            try {
                final DdiRestApi client = new ClientBuilder()
                        .withBaseUrl(url)
                        .withGatewayToken(gatewayToken)
                        .withTargetToken(targetToken)
                        .withHttpBuilder(buildOkHttpClient())
                        .build();
                ufService = UFService.builder()
                        .withClient(client)
                        .withRetryDelayOnCommunicationError(delay)
                        .withTenant(tenant)
                        .withControllerId(controllerId)
                        .withInitialState(initialState)
                        .withTargetData(()->args)
                        .build();

                ufService.addObserver(new ObserverState(apiMode));
                startStopService(true);
                if (initialState.getStateName() == State.StateName.UPDATE_STARTED) {
                    ufService.setUpdateSucceffullyUpdate(UpdateSystem.successInstallation());
                }
            }catch (IllegalStateException | IllegalArgumentException e){
                sharedPreferences.edit().putBoolean(sharedPreferencesServiceEnableKey, false).apply();
                Toast.makeText(this,"Update Factory configuration error",Toast.LENGTH_LONG)
                        .show();
            }
        }
    }

    private ArrayList<Messenger> mClients = new ArrayList<Messenger>();

    private UFServiceConfiguration getCurrentConfiguration(SharedPreferencesWithObject sharedPreferences){
        final boolean serviceIsEnable = ufService != null && sharedPreferences.getBoolean(sharedPreferencesServiceEnableKey, false);
        final String url = sharedPreferences.getString(sharedPreferencesServerUrlKey, "");
        final String controllerId = sharedPreferences.getString(sharedPreferencesControllerIdKey, "");
        final String gatewayToken = sharedPreferences.getString(sharedPreferencesGatewayToken, "");
        final String targetToken = sharedPreferences.getString(sharedPreferencesTargetToken, "");
        final String tenant = sharedPreferences.getString(sharedPreferencesTenantKey, "");
        final long delay = sharedPreferences.getLong(sharedPreferencesRetryDelayKey, 30000);
        final boolean apiMode = sharedPreferences.getBoolean(sharedPreferencesApiModeKey, true);
        final Map<String,String> args = sharedPreferences.getObject(sharedPreferencesArgs, new HashMap<String,String>().getClass());
        return UFServiceConfiguration.builder()
                .witArgs(args)
                .witEnable(serviceIsEnable)
                .withApiMode(apiMode)
                .withControllerId(controllerId)
                .withGetawayToken(gatewayToken)
                .withRetryDelay(delay)
                .withTargetToken(targetToken)
                .withTenant(tenant)
                .withUrl(url)
                .build();
    }
    private class IncomingHandler extends Handler {
        @Override
        public void handleMessage(Message msg) {
            switch (msg.what) {
                case MSG_CONFIGURE_SERVICE:
                    final UFServiceConfiguration configuration = (UFServiceConfiguration) msg.getData().getSerializable(SERVICE_DATA_KEY);
                    saveServiceConfigurationToSharedPreferences(configuration);
                    buildServiceFromPreferences(true);
                    break;
                case MSG_REGISTER_CLIENT:
                    mClients.add(msg.replyTo);
                    break;
                case MSG_UNREGISTER_CLIENT:
                    mClients.remove(msg.replyTo);
                    break;
                case MSG_AUTHORIZATION_RESPONSE:
                    ufService.setAuthorized(msg.getData().getBoolean(SERVICE_DATA_KEY));
                    break;
                case MSG_RESUME_SUSPEND_UPGRADE:
                    ufService.restartSuspendState();
                    break;
                case MSG_SYNCH_REQUEST:
                    final SharedPreferencesWithObject sharedPreferences = getSharedPreferences(sharedPreferencesFile, MODE_PRIVATE);
                    UpdateFactoryService.this.sendMessage(getCurrentConfiguration(sharedPreferences), MSG_SERVICE_CONFIGURATION_STATUS, msg.replyTo);
                    if(ufService == null){
                        return;
                    }
                    final UFServiceMessage lastMessage = sharedPreferences.getObject(SHARED_PREFERENCES_LAST_NOTIFY_MESSAGE, UFServiceMessage.class);
                    if(lastMessage != null){
                        UpdateFactoryService.this.sendMessage(lastMessage, MSG_SEND_STRING, msg.replyTo);
                    }
                    State lastState = sharedPreferences.getObject(sharedPreferencesCurrentStateKey, State.class);
                    if(lastState.getStateName() == State.StateName.AUTHORIZATION_WAITING){
                        UpdateFactoryService.this.sendMessage(((State.AuthorizationWaitingState)lastState).getState().getStateName().name(),
                                MSG_AUTHORIZATION_REQUEST,
                                msg.replyTo);
                    }
                    break;
                default:
                    super.handleMessage(msg);
            }
        }
    }

    private void saveServiceConfigurationToSharedPreferences(UFServiceConfiguration configuration) {
        SharedPreferencesWithObject sharedPreferences;
        sharedPreferences = getSharedPreferences(sharedPreferencesFile, MODE_PRIVATE);
        SharedPreferences.Editor editor = sharedPreferences.edit();
        editor.putString(sharedPreferencesControllerIdKey, configuration.getControllerId());
        editor.putString(sharedPreferencesTenantKey, configuration.getTenant());
        editor.putString(sharedPreferencesServerUrlKey, configuration.getUrl());
        editor.putString(sharedPreferencesGatewayToken, configuration.getGatewayToken());
        editor.putString(sharedPreferencesTargetToken, configuration.getTargetToken());
        editor.putLong(sharedPreferencesRetryDelayKey, configuration.getRetryDelay());
        editor.putBoolean(sharedPreferencesApiModeKey, configuration.isApiMode());
        editor.putBoolean(sharedPreferencesServiceEnableKey, configuration.isEnalbe());
        editor.apply();
        sharedPreferences.putAndCommitObject(sharedPreferencesArgs,configuration.getArgs());
    }

    private void startStopService(boolean serviceIsEnable) {
        if(ufService == null){
            return;
        }
        if(serviceIsEnable) {
            ufService.start();
        } else {
            ufService.stop();
        }
    }

    private void sendMessage(Serializable messageContent, int code, Messenger messenger) {
        final Message message = getMessage(messageContent, code);
        try {
            messenger.send(message);
        } catch (RemoteException e) {
            e.printStackTrace();
        }
    }

    private void sendMessage(Serializable messageContent, int code){
        final Message message = getMessage(messageContent, code);
        int i = 0;
        while (i<mClients.size()) {
            try {
                mClients.get(i).send(message);
                i++;
            } catch (RemoteException e) {
                mClients.remove(i);
            }
        }
    }

    @NonNull
    private Message getMessage(Serializable messageContent, int code) {
        final Message message = Message.obtain(null, code);
        final Bundle data = new Bundle();
        data.putSerializable(SERVICE_DATA_KEY, messageContent);
        message.setData(data);
        return message;
    }

    private final Messenger mMessenger = new Messenger(new IncomingHandler());

    @Override
    public IBinder onBind(Intent intent) {
        return mMessenger.getBinder();
    }

    private class ObserverState implements Observer {
        private final boolean apiMode;

        public ObserverState(boolean apiMode) {
            this.apiMode = apiMode;
        }

        @Override
        public void update(Observable observable, Object o) {
            if(o instanceof UFService.SharedEvent) {
                final UFService.SharedEvent eventNotify = (UFService.SharedEvent) o;
                final Event event = eventNotify.getEvent();
                final State newState = eventNotify.getNewState();

                final UFServiceMessage message = new UFServiceMessage(
                        event.getEventName().name(),
                        eventNotify.getOldState().getStateName().name(),
                        newState.getStateName().name(),
                        getSuspend(newState)
                );
                writeObjectToSharedPreference(message, SHARED_PREFERENCES_LAST_NOTIFY_MESSAGE);
                sendMessage(message, MSG_SEND_STRING);
                writeObjectToSharedPreference(eventNotify.getNewState(), sharedPreferencesCurrentStateKey);
                switch (newState.getStateName()){
                    case AUTHORIZATION_WAITING:
                        final State.StateName auth = ((State.AuthorizationWaitingState) newState).getState().getStateName();
                        if(apiMode){
                            sendMessage(auth.name(), MSG_AUTHORIZATION_REQUEST);
                        }else {
                            showAuthorizationDialog(auth);
                        }
                        break;
                    case SAVING_FILE:
                        final State.SavingFileState savingFileState = ((State.SavingFileState) newState);
                        UpdateSystem.copyFile(savingFileState.getInputStream());
                        break;
                    case UPDATE_STARTED:
                        if(UpdateSystem.verify()){
                            UpdateSystem.install(getApplicationContext(), ((State.StateWithAction)newState).getActionId());
                        } else {
                            ufService.setUpdateSucceffullyUpdate(false);
                            Toast.makeText(getApplicationContext(),getString(R.string.invalid_update), Toast.LENGTH_LONG).show();
                        }
                        break;
                }
            }
        }
    }

    private void showAuthorizationDialog(State.StateName auth) {
        final Intent intent = new Intent(UpdateFactoryService.this, MainActivity.class);
        intent.putExtra(MainActivity.INTENT_TYPE_EXTRA_VARIABLE, auth == State.StateName.UPDATE_DOWNLOAD ?
                MainActivity.INTENT_TYPE_EXTRA_VALUE_DOWNLOAD : MainActivity.INTENT_TYPE_EXTRA_VALUE_REBOOT);
        intent.addFlags(FLAG_ACTIVITY_NEW_TASK);
        startActivity(intent);
    }

    private Suspend getSuspend(State state){
        if(state.getStateName() != State.StateName.WAITING){
            return NONE;
        }
        State.WaitingState waitingState = (State.WaitingState) state;
        if(!waitingState.hasInnerState()){
            return NONE;
        }
        return waitingState.getState().getStateName() == State.StateName.UPDATE_DOWNLOAD ?
                DOWNLOAD : UPDATE;
    }

    @Override
    public SharedPreferencesWithObject getSharedPreferences(String name, int mode){
        return new SharedPreferencesWithObject(super.getSharedPreferences(name, mode));
    }

    private void writeObjectToSharedPreference(Serializable obj, String key){
        final SharedPreferencesWithObject sharedPreferences = getSharedPreferences(sharedPreferencesFile,MODE_PRIVATE);
        sharedPreferences.putAndCommitObject(key, obj);
    }

    private void initSharedPreferencesKeys(){
        sharedPreferencesFile = getString(R.string.shared_preferences_file);
        sharedPreferencesCurrentStateKey = getString(R.string.shared_preferences_current_state_key);
        sharedPreferencesServerUrlKey = getString(R.string.shared_preferences_server_url_key);
        sharedPreferencesApiModeKey = getString(R.string.shared_preferences_api_mode_key);
        sharedPreferencesTenantKey = getString(R.string.shared_preferences_tenant_key);
        sharedPreferencesControllerIdKey = getString(R.string.shared_preferences_controller_id_key);
        sharedPreferencesRetryDelayKey = getString(R.string.shared_preferences_retry_delay_key);
        sharedPreferencesServiceEnableKey = getString(R.string.shared_preferences_is_enable_key);
        sharedPreferencesGatewayToken = getString(R.string.shared_preferences_gateway_token_key);
        sharedPreferencesTargetToken = getString(R.string.shared_preferences_target_token_key);
        sharedPreferencesArgs = getString(R.string.shared_preferences_args_key);
    }

    private static UpdateFactoryServiceCommand ufServiceCommand;
    private UFService ufService;
    private String sharedPreferencesCurrentStateKey;
    private String sharedPreferencesServerUrlKey;
    private String sharedPreferencesApiModeKey;
    private String sharedPreferencesServiceEnableKey;
    private String sharedPreferencesTenantKey;
    private String sharedPreferencesControllerIdKey;
    private String sharedPreferencesRetryDelayKey;
    private String sharedPreferencesGatewayToken;
    private String sharedPreferencesTargetToken;
    private String sharedPreferencesFile;
    private String sharedPreferencesArgs;

    private static final String SHARED_PREFERENCES_LAST_NOTIFY_MESSAGE = "LAST_NOTIFY_MESSAGE";

}

File: uf-client-service/src/main/kotlin/com/kynetics/uf/android/UpdateFactoryService.kt
/*
 * Copyright © 2017-2023  Kynetics  LLC
 * 
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 */

package com.kynetics.uf.android

import android.app.Notification
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.*
import android.util.Log
import androidx.core.app.NotificationCompat
import com.kynetics.uf.android.apicomptibility.ApiVersion
import com.kynetics.uf.android.client.RestartableClientService
import com.kynetics.uf.android.communication.CommunicationApi
import com.kynetics.uf.android.communication.CommunicationApiStrategy
import com.kynetics.uf.android.communication.messenger.MessageHandler
import com.kynetics.uf.android.communication.messenger.MessengerHandler
import com.kynetics.uf.android.configuration.AndroidDeploymentPermitProvider
import com.kynetics.uf.android.configuration.AndroidMessageListener
import com.kynetics.uf.android.configuration.ConfigurationHandler
import com.kynetics.uf.android.content.SharedPreferencesWithObject
import com.kynetics.uf.android.content.SharedPreferencesFactory
import com.kynetics.uf.android.ui.MainActivity
import com.kynetics.uf.android.update.CurrentUpdateState
import com.kynetics.uf.android.update.system.SystemUpdateType
import de.psdev.slf4j.android.logger.AndroidLoggerAdapter
import de.psdev.slf4j.android.logger.LogLevel
import org.eclipse.hara.ddiclient.api.MessageListener
import java.lang.ref.WeakReference

/*
 * @author Daniele Sergio
 */
class UpdateFactoryService : Service(), UpdateFactoryServiceCommand {

    override fun authorizationGranted() {
        softDeploymentPermitProvider?.allow(true)
    }

    override fun authorizationDenied() {
        softDeploymentPermitProvider?.allow(false)
    }

    private val mMessenger = Messenger(IncomingHandler(this))

    private var mNotificationManager: NotificationManager? = null
    private var systemUpdateType: SystemUpdateType = SystemUpdateType.SINGLE_COPY

    var softDeploymentPermitProvider: AndroidDeploymentPermitProvider? = null
    private var messageListener: MessageListener? = null

    private var sharedPreferencesFile: String? = null
    private var configurationHandler: ConfigurationHandler? = null
    private var ufService: RestartableClientService? = null

    private val api: CommunicationApi by lazy {
        CommunicationApiStrategy.newInstance(
            configurationHandler!!,
            ufService!!,
            softDeploymentPermitProvider!!
        )
    }
    override fun configureService() {
        if(ufService == null){
            ufService = RestartableClientService.newInstance(softDeploymentPermitProvider!!, listOf(messageListener!!))
        }
        when {
            configurationHandler!=null -> {
                ufService?.restartService(configurationHandler!!)
            }
            else -> {
                Log.w(TAG, "Service cant be configured because configuration handler is null")
            }
        }
    }

    private lateinit var forcePingPendingIntent: PendingIntent
    lateinit var currentUpdateState: CurrentUpdateState

    override fun onCreate() {
        super.onCreate()
        AndroidLoggerAdapter.setLogTag("Update Factory Client")
        if(BuildConfig.DEBUG){
            AndroidLoggerAdapter.setLogLevel(LogLevel.TRACE)
        }
        mNotificationManager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        sharedPreferencesFile = getString(R.string.shared_preferences_file)
        configurationHandler = ConfigurationHandler( this, getSharedPreferences(sharedPreferencesFile, Context.MODE_PRIVATE))
        systemUpdateType = SystemUpdateType.getSystemUpdateType()
        ufServiceCommand = this
        softDeploymentPermitProvider = AndroidDeploymentPermitProvider.build(configurationHandler!!, mNotificationManager!!, this)
        messageListener = AndroidMessageListener(this)
        currentUpdateState = CurrentUpdateState(this)

        val forcePingIntent = Intent(FORCE_PING_ACTION)

        forcePingPendingIntent = PendingIntent.getBroadcast(this, 1, forcePingIntent, 0)

        // add actions here !
        val intentFilter = IntentFilter()
        intentFilter.addAction(FORCE_PING_ACTION)

        val receiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
                if (intent.action == FORCE_PING_ACTION) {
                    ufService?.forcePing()
                    MessengerHandler.onAction(MessageHandler.Action.FORCE_PING)
                    val closeIntent = Intent(Intent.ACTION_CLOSE_SYSTEM_DIALOGS)
                    sendBroadcast(closeIntent)
                }
            }
        }

        registerReceiver(receiver, intentFilter)
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        Log.i(TAG, String.format("service's starting with version %s (%s)", BuildConfig.VERSION_NAME, BuildConfig.VERSION_CODE))
        startForeground()
        isRunning = true
        var serviceConfiguration = configurationHandler?.getConfigurationFromFile()
        if (serviceConfiguration == null && intent != null) {
            serviceConfiguration = configurationHandler?.getServiceConfigurationFromIntent(intent)
        } else if (serviceConfiguration != null) {
            Log.i(TAG, "Loaded new configuration from file")
        }
        if (configurationHandler?.getCurrentConfiguration() != serviceConfiguration) {
            configurationHandler?.saveServiceConfigurationToSharedPreferences(serviceConfiguration)
        }
        configureService()
        return START_STICKY
    }

    // todo add api to configure targetAttibutes (separete  d from serviceConfiguration)
    private class IncomingHandler(service: UpdateFactoryService) : Handler(Looper.myLooper()!!) {
        private val updateFactoryServiceRef = WeakReference(service)

        private fun <T> WeakReference<T>.execute(action: T.() -> Unit){
            get()?.let { ref ->
                ref.action()
            }
        }
        override fun handleMessage(msg: Message) {
            updateFactoryServiceRef.execute { api.onMessage(msg) }
        }

    }

    override fun onBind(intent: Intent): IBinder? {
        startService(this)
        return mMessenger.binder
    }

    override fun getSharedPreferences(name: String?, mode: Int): SharedPreferencesWithObject {
        return SharedPreferencesFactory.get(applicationContext, name, mode)
    }

    private fun startForeground() {
        if(mNotificationManager != null) {
            ApiVersion.fromVersionCode().configureChannel(CHANNEL_ID, getString(R.string.app_name), mNotificationManager!!)
            startForeground(NOTIFICATION_ID, getNotification(""))
        }
    }

    fun getNotification(notificationContent: String, forcePingAction: Boolean = false): Notification {

        val notificationBuilder = NotificationCompat.Builder(this, CHANNEL_ID)
                .setSmallIcon(R.drawable.uf_logo)
                .setStyle(NotificationCompat.BigTextStyle().bigText(notificationContent))
                .setContentTitle(getString(R.string.update_factory_notification_title))
                .setContentText(notificationContent)
                .setPriority(NotificationCompat.PRIORITY_DEFAULT)
        if (forcePingAction) {
            notificationBuilder.addAction(android.R.drawable.ic_popup_sync, "Grant Authorization", forcePingPendingIntent)
        }
        return notificationBuilder.build()
    }

    companion object {
        enum class AuthorizationType(
            val toActionOnGranted: MessageHandler.Action,
            val toActionOnDenied: MessageHandler.Action
        ) {
            DOWNLOAD(
                    MessageHandler.Action.AUTH_DOWNLOAD_GRANTED,
                    MessageHandler.Action.AUTH_DOWNLOAD_DENIED
            ) {
                override val extra = MainActivity.INTENT_TYPE_EXTRA_VALUE_DOWNLOAD
                override val event = "WaitingDownloadAuthorization"
            },

            UPDATE(
                MessageHandler.Action.AUTH_UPDATE_GRANTED,
                    MessageHandler.Action.AUTH_UPDATE_DENIED) {
                override val extra: Int = MainActivity.INTENT_TYPE_EXTRA_VALUE_REBOOT
                override val event = "WaitingUpdateAuthorization"
            };

            abstract val extra: Int
            abstract val event: String
        }

        @JvmStatic
        fun startService(context: Context) {
            if (!isRunning) {
                val myIntent = Intent(context, UpdateFactoryService::class.java)
                ApiVersion.fromVersionCode().startService(context, myIntent)
            }
        }

        var isRunning = false

        @JvmStatic
        var ufServiceCommand: UpdateFactoryServiceCommand? = null
        private const val FORCE_PING_ACTION = "ForcePing"
        private const val CHANNEL_ID = "UPDATE_FACTORY_NOTIFICATION_CHANNEL_ID"
        const val NOTIFICATION_ID = 1
        private val TAG = UpdateFactoryService::class.java.simpleName
    }
}
